source/common/upstream/conn_pool_map.h

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include <map>
+#include <vector>
+
+#include "envoy/event/dispatcher.h"
+
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Upstream {
+/**
+ *  A class mapping keys to connection pools, with some recycling logic built in.
+ */
+template <typename KEY_TYPE, typename POOL_TYPE> class ConnPoolMap {
+public:
+  using PoolFactory = std::function<std::unique_ptr<POOL_TYPE>()>;
+  using DrainedCb = std::function<void()>;
+
+  ConnPoolMap(Event::Dispatcher& dispatcher);
+  ~ConnPoolMap();
+  /**
+   * Returns an existing pool for `key`, or creates a new one using `factory`. Note that it is
+   * possible for this to fail if a limit on the number of pools allowed is reached.
+   * @return The pool corresponding to `key`, or `absl::nullopt`
+   */
+  absl::optional<POOL_TYPE*> getPool(KEY_TYPE key, PoolFactory factory);
+
+  /**
+   * @return the number of pools
+   */
+  size_t size() const;
+
+  /**
+   * Destroys all mapped pools.
+   */
+  void clear();
+
+  /**
+   * Adds a drain callback to all mapped pools. Any future mapped pools with have the callback
+   * automatically added. Be careful with the callback. If it itself calls into `this`, modifying
```

Comment:
Are there any asserts we can put in to prevent people from doing this bad behavior?

Reply:
Hmmm. It's actually similar to a mutex -- we really don't want to recursively enter anything within this class. The simple way to do this is to probably just set a boolean on entry to any function, and assert that it is not set first. I worry a bit about exceptions here... Do we have any helper classes/utilities providing "finally"-like behaviour I could use?

Reply:
I ended up writing a simple pair of classes to do this. One owns a bool. If it is set it is "entered", it trips the assert. It unsets it on "exit". Another is an RAII-like object to wrap this. Just declare it at the top of your function, and you're good to go. Seems to work well enough.